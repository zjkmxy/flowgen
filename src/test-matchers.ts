import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
