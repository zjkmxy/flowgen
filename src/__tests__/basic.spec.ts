import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle basic keywords", async () => {
  const ts = `type A = {
    a: void,
    b: string,
    c: any,
    d: number,
    e: boolean,
    f: null,
    g: undefined,
    h: object,
    i: 1,
    j: 2,
    k: true,
    l: false,
    m: 'foo',
    n: 'bar',
    o: never,
    p: unknown,
    r: -1,
    s: -2,
    t: symbol,
    u: unique symbol,
    v: readonly [1, 2, 3],
    w: readonly string[],
    x: RegExpMatchArray
  }`;

  {
    const result = compiler.compileDefinitionString(ts, {
      quiet: true,
      inexact: true,
    });
    expect(await beautify(result)).toMatchSnapshot();
    expect(result).toBeValidFlowTypeDeclarations();
  }

  {
    const result = compiler.compileDefinitionString(ts, {
      quiet: true,
      inexact: false,
    });
    expect(await beautify(result)).toMatchSnapshot();
    expect(result).toBeValidFlowTypeDeclarations();
  }
});

it("should handle class types", async () => {
  const ts = `
  declare export class Foo {
  }`;

  {
    const result = compiler.compileDefinitionString(ts, { quiet: true });
    expect(await beautify(result)).toMatchSnapshot();
    expect(result).toBeValidFlowTypeDeclarations();
  }

  {
    const result = compiler.compileDefinitionString(ts, {
      quiet: true,
      inexact: false,
    });
    expect(await beautify(result)).toMatchSnapshot();
    expect(result).toBeValidFlowTypeDeclarations();
  }
});
