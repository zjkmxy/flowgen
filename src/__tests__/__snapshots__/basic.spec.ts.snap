// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle basic keywords 1`] = `
"declare type A = {
  a: void,
  b: string,
  c: any,
  d: number,
  e: boolean,
  f: null,
  g: void,
  h: { [key: string]: any },
  i: 1,
  j: 2,
  k: true,
  l: false,
  m: "foo",
  n: "bar",
  o: empty,
  p: mixed,
  r: -1,
  s: -2,
  t: Symbol,
  u: Symbol,
  v: [1, 2, 3],
  w: $ReadOnlyArray<string>,
  x: RegExp$matchResult,
  ...
};
"
`;

exports[`should handle basic keywords 2`] = `
"declare type A = {
  a: void,
  b: string,
  c: any,
  d: number,
  e: boolean,
  f: null,
  g: void,
  h: { [key: string]: any },
  i: 1,
  j: 2,
  k: true,
  l: false,
  m: "foo",
  n: "bar",
  o: empty,
  p: mixed,
  r: -1,
  s: -2,
  t: Symbol,
  u: Symbol,
  v: [1, 2, 3],
  w: $ReadOnlyArray<string>,
  x: RegExp$matchResult,
};
"
`;

exports[`should handle class types 1`] = `
"declare export class Foo {}
"
`;

exports[`should handle class types 2`] = `
"declare export class Foo {}
"
`;
